lib/OpenQA/Worker/Cache.pm

```diff
@@ -372,144 +285,114 @@ sub _update_asset_last_use {
     }
     else {
         log_info "CACHE: updating the $asset with $etag and $size";
     }
 
-    return !!0 if !defined $res || $res == 0 || $@;
-    return !!1 if defined $res && $res == 1;
+    my $result = $self->dbh->db->select('assets', [qw(etag size last_use filename)], {filename => $asset})->arrays;
+
+    return !!0 if $result->size == 0 || $@;
+    $self->increase($size) and return !!1 if $result->size == 1;
 }
 
 sub purge_asset {
     my ($self, $asset) = @_;
-    my $sql = "DELETE FROM assets WHERE filename = ?";
-    my $dbh = $self->dbh;
-
     eval {
-        $dbh->do("BEGIN EXCLUSIVE");
-        $dbh->prepare($sql)->execute($asset) or die $dbh->errstr;
-        $dbh->commit or die $dbh->errstr;
-        unlink($asset) or eval { log_error "CACHE: Could not remove $asset" if -e $asset };
-        log_debug "CACHE: removed $asset";
+        my $tx = $self->dbh->db->begin();
+        $self->dbh->db->delete('assets', {filename => $asset});
+        $tx->commit;
+        if (-e $asset) {
+            unlink($asset) or eval { log_error "CACHE: Could not remove $asset" if -e $asset };
+            log_debug "CACHE: removed $asset";
+        }
+        else {
+            log_debug "CACHE: requested to remove unnexistant asset $asset";
+        }
     };
 
     if ($@) {
-        log_error "purge_asset: Rolling back $@";
-        eval { $dbh->rollback };
-        log_error "Rolling back failed: $@";
+        log_error "purge_asset: $@";
         return !!0;
     }
     return !!1;
 }
 
-sub cache_cleanup {
+sub file_size { (stat(pop))[7] }
+
+sub cache_sync {
     my $self     = shift;
     my $location = $self->location;
-    my @assets
-      = `find $location -maxdepth 1 -type f -name '*.img' -o -name '*.qcow2' -o -name '*.iso' -o -name '*.vhd' -o -name '*.vhdx'`;
+    my $ext;
+    $ext .= "-o -name '*.$_' " for qw(qcow2 iso vhd vhdx);
+    my @assets = `find $location -maxdepth 2 -type f -name '*.img' $ext`;
     chomp @assets;
+    $self->cache_real_size(0);
     foreach my $file (@assets) {
-        my $asset_size = (stat $file)[7];
+        my $asset_size = $self->file_size($file);
         next if !defined $asset_size;
-        $self->cache_real_size($self->cache_real_size + $asset_size) if $self->asset_lookup($file);
+        $self->increase($asset_size) if $self->asset_lookup($file);
     }
 }
 
 sub asset_lookup {
     my ($self, $asset) = @_;
     my $sth;
-    my $sql;
-    my $lock_granted;
     my $result;
-    my $dbh = $self->dbh;
-
     eval {
-        $dbh->do("BEGIN EXCLUSIVE");
-        $sql    = "SELECT filename, etag, last_use, size from assets where filename = ?";
-        $sth    = $dbh->prepare($sql);
-        $result = $dbh->selectrow_hashref($sql, undef, $asset);
-        $dbh->commit or die $dbh->errstr;
+        my $tx = $self->dbh->db->begin('exclusive');
+        $result = $self->dbh->db->select('assets', [qw(filename etag last_use size)], {filename => $asset});
+        $tx->commit;
```

Comment:
Another case where you need to hold on to the db handle instead of calling `$self->dbh->db` repeatedly.